### PR TITLE
feat: UMW Roles

### PR DESCRIPTION
**Pull request by tomerlichtash**
## Related Issues

Fixes <link_to_github_issue>

Related issue: https://github.com/descope/etc/issues/5601

## Related PRs

| branch       | PR         |
| ------------ | ---------- |
| service a PR | Link to PR |
| service b PR | Link to PR |

## Description

A few sentences describing the overall goals of the pull request's commits.

## Must

- [x] Tests
- [ ] Documentation (if applicable)

### CODE REVIEW

File: packages/user-management-widget/src/lib/widget/drivers/MultiSelectDriver.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { BaseDriver } from './BaseDriver';
+
+export class MultiSelectDriver extends BaseDriver {
+  nodeName = 'descope-combo-box';
+  // nodeName = 'descope-button-multi-selection-group';
```
Comment (tomerlichtash) [inline packages/user-management-widget/src/lib/widget/drivers/MultiSelectDriver.ts:5]: Replace with MultiSelect component

File: packages/user-management-widget/src/lib/widget/mockTemplates/createUserTemplate.ts [hunk 2]
```diff
@@ -9,6 +9,8 @@ export default `
   st-vertical-padding="0rem"
   st-border-radius="sm"
 >
+  <descope-combo-box bordered="true" data-id="roles-multiselect"></descope-combo-box>
```
Comment (tomerlichtash) [inline packages/user-management-widget/src/lib/widget/mockTemplates/createUserTemplate.ts:12]: Replace with MultiSelect component

Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/mockTemplates/createUserTemplate.ts:12]: And move to place

File: packages/user-management-widget/src/lib/widget/api/apiPaths.ts [hunk 3]
```diff
@@ -5,5 +5,6 @@ export const apiPaths = {
     create: '/v1/mgmt/user/create',
     expirePassword: '/v1/mgmt/user/password/expire',
     customAttributes: '/v1/mgmt/user/customattributes',
+    allRoles: '/v1/mgmt/role/all',
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/api/apiPaths.ts:8]: I don't think this should be under user

File: packages/user-management-widget/src/lib/widget/api/sdk/createUserSdk.ts [hunk 4]
```diff
@@ -140,11 +140,22 @@ export const createUserSdk = ({
     return res.json();
   };
 
+  const getTenantRoles = async () => {
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/api/sdk/createUserSdk.ts:143]: Also this one, need to put it under a different name

File: packages/user-management-widget/src/lib/widget/mockTemplates/createUserTemplate.ts (hunk 2 above)
Comment (nirgur) [reply packages/user-management-widget/src/lib/widget/mockTemplates/createUserTemplate.ts:12]: This should be added to the page schema after we are adding the multi-select to the flow components

File: packages/user-management-widget/src/lib/widget/state/asyncActions/getTenantRoles.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  {
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    onFulfilled: (state, action) => {
+      // eslint-disable-next-line no-param-reassign
+      state.tenantRoles.data = action.payload.roles;
+    },
+  },
+  withRequestStatus((state: State) => state.customAttributes),
+  // eslint-disable-next-line @typescript-eslint/no-shadow
+  withRequestStatus((state) => state.customAttributes),
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/state/asyncActions/getTenantRoles.ts:22]: Why do we have it twice?
And also, we are not passing the correct state section

File: packages/user-management-widget/src/lib/widget/state/selectors.ts [hunk 6]
```diff
@@ -12,9 +13,18 @@ export const getUsersList = createSelector(getRawUsersList, (users) =>
...
     status: userStatusMappings[user.status] || user.status,
+    roles: user.roleNames
+      .concat(user.userTenants?.[0]?.roleNames)
+      .filter(Boolean)
+      .sort()
+      .join(', '),
   })),
 );
 
+export const getTenantRoles = createSelector(getRawTenantRoles, (roles) =>
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/state/selectors.ts:24]: I'm not sure if this is the right place to transform the data to match the component structure
I would expect it to happen on the component init fn
WDYT?

File: packages/user-management-widget/src/lib/widget/state/selectors.ts [hunk 7]
```diff
@@ -12,9 +13,18 @@ export const getUsersList = createSelector(getRawUsersList, (users) =>
   users.map((user) => ({
     ...user,
     status: userStatusMappings[user.status] || user.status,
+    roles: user.roleNames
+      .concat(user.userTenants?.[0]?.roleNames)
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/state/selectors.ts:17]: We need the tenant ID, using 0 index is problematic

File: packages/user-management-widget/src/lib/widget/state/selectors.ts [hunk 8]
```diff
@@ -12,9 +13,18 @@ export const getUsersList = createSelector(getRawUsersList, (users) =>
...
+      .filter(Boolean)
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/state/selectors.ts:18]: I think you can concat an empty array if there is no roleNames and eliminate the filter

File: packages/user-management-widget/src/lib/widget/state/selectors.ts [hunk 9]
```diff
@@ -12,9 +13,18 @@ export const getUsersList = createSelector(getRawUsersList, (users) =>
...
+      .sort()
+      .join(', '),
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/state/selectors.ts:20]: I don't think we should join it here, or join it at all
AFAIR, we can pass an array to the grid and it should be ok

**Review by nirgur — COMMENTED** [review]
Looks good
Added some comments

File: packages/user-management-widget/src/lib/widget/api/apiPaths.ts (hunk 3 above)
Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/api/apiPaths.ts:8]: Fixed

File: packages/user-management-widget/src/lib/widget/state/selectors.ts (hunk 7 above)
Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/state/selectors.ts:17]: Agreed. Related to pending issue that we need to resolve: https://github.com/descope/etc/issues/5601

File: packages/user-management-widget/src/lib/widget/state/asyncActions/getTenantRoles.ts (hunk 5 above)
Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/state/asyncActions/getTenantRoles.ts:22]: Copy-pasted from `getCustomAttributes` asyncActions; I'll remove it there as well.

File: packages/user-management-widget/src/lib/widget/state/selectors.ts (hunk 6 above)
Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/state/selectors.ts:24]: Moved to component init

File: packages/user-management-widget/src/lib/widget/api/sdk/createUserSdk.ts (hunk 4 above)
Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/api/sdk/createUserSdk.ts:143]: Fixed

File: packages/user-management-widget/src/lib/widget/drivers/MultiSelectDriver.ts [hunk 10]
```diff
@@ -4,7 +4,8 @@ export class MultiSelectDriver extends BaseDriver {
   nodeName = 'descope-multi-select-combo-box';
 
   setData(data: { label: string; value: string }[]) {
-    this.ele?.setAttribute('data', JSON.stringify(data.toSorted()));
+    const sortedRoles = data;
```
Comment (nirgur) [inline packages/user-management-widget/src/lib/widget/drivers/MultiSelectDriver.ts:7]: why do we need the `sortedRoles`?

Comment (tomerlichtash) [reply packages/user-management-widget/src/lib/widget/drivers/MultiSelectDriver.ts]: Removed

**Review by nirgur — APPROVED** [review]

**Pull request merged**